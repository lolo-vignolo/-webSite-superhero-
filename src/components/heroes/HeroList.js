import React, { useMemo } from 'react'
import { getHeroByPublisher } from '../../selectors/getHeroByPublisher'
import { HeroCard } from './HeroCard';

//publisher viene del file heroes.
export const HeroList = ({publisher}) => {

    const heroes = useMemo(() =>  getHeroByPublisher(publisher), [publisher]);

    

    return (
        
            <div className="card columns animate__animated animate__fadeIn">
        
            {heroes.map( hero => (
                    <HeroCard key= {hero.id}
                            {...hero}
                            //   superhero=  {hero.superhero} //podrÃ­a hacerlo asÃ­ con cada uno de los props, pero usando los tres puntos que como si lo hago con todos.
                    />
                ))
            }
        </div>
        
    )
}
