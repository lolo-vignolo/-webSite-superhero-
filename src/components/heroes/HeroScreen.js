import { useMemo } from 'react';
import { useParams, Navigate, useNavigate } from 'react-router-dom'
import { getHeroById as getHeroesById } from '../../selectors/getHeroById';


export const HeroScreen = () => {

    const { heroeId } = useParams();
    const navigate = useNavigate()

    const hero = useMemo( () => getHeroesById(heroeId), [ heroeId ]);
    
    console.log(hero);

    const handleReturn = () => {
        navigate( -1 );
    }


    if (!hero) {
        return <Navigate to='/' />
    }
    
    const {
        id,
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    console.log(id);
    const imagePath = `/assets/${ id }.jpg`;
   
    return (
        <div className="row mt-5">
            <div className="col-4">
                <img 
                    src={ imagePath } 
                    alt={ superhero }
                    className="img-thumbnail animate__animated animate__fadeInLeft"
                />
            </div>

            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{ hero.superhero }</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"> <b>Alter ego:</b> { alter_ego } </li>
                    <li className="list-group-item"> <b>Publisher:</b> { publisher } </li>
                    <li className="list-group-item"> <b>First Appearance:</b> { first_appearance } </li>
                </ul>

                <h5 className="mt-3">Characters</h5>
                <p>{ characters }</p>

                <button
                    className="btn btn-outline-info"
                    onClick={ handleReturn }
                >
                    Regresar 
                </button>

            </div>

        </div>
    )
}



// import React, { useMemo } from 'react';
// import { Redirect, useParams } from 'react-router';
// import { getHeroById } from '../../selectors/getHeroById';

// export const HeroScreen = ({history}) => {

//     // este useParam es un customHook de react que sirve para obtener informacion de esa lista de 
//     // items que encuentro en el components de mi sitio web. heroId es uno de esos items.
//     //useParam va a extraer los parametros que vayan por el URL.

//     const {heroId} = useParams();
   
    

//     const hero = useMemo(() =>  getHeroById(heroId), [heroId])

//     console.log(hero);

//     if (!hero){
//         return <Redirect to="/" />;
//     }

//     const handleReturn = ()=>{
//         if(history.length <= 2){
//             history.push("/");
//         }else{
//             history.goBack();
//         }

//     // this history.goBack() is inside of HeroScreen in components in the console. There IÂ´ll find the different acctions 
//     //that I can carry out with my button. 
       
        
//     } 

//     const {
//         superhero,
//         publisher,
//         alter_ego,
//         first_appearance,
//         characters,
//     } = hero
    
   


//     return (
//         <div className="row mt-5">
//             <div className = "col-4"> 
//                 <img
//                 src={`../assets/heroes/${heroId}.jpg`}
//                 alt = {superhero}
//                 className = "img-thumbnail animate__animated animate__fadeInLeft"
//                 />
             
//             </div>
//             <div className = "col-8">
//             <h3>{superhero}</h3>
//             <ul className= "list-group list-group-flush">
//                 <li className = "list-group-item"><b>Alter ego:</b> {alter_ego}</li>
//                 <li className = "list-group-item"><b>publisher:</b> {publisher}</li>
//                 <li className = "list-group-item"><b>first appearance:</b> {first_appearance}</li>
//             </ul>

//             <h5> characters: </h5>
//             <p>{characters} </p>

//             <button 
//                 className="btn btn-outline-info"
//                 onClick = {handleReturn}
//                 >
//                 Return

//             </button>


//             </div>

//         </div>
//     )
// }
