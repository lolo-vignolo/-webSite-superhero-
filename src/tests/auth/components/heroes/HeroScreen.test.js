import { mount } from "enzyme"
import { MemoryRouter, Route, Router } from "react-router"
import { HeroScreen } from "../../../../components/heroes/HeroScreen"

describe('test component <HeroScreen />', () => {
    


    test('should shows itself correctly ', () => {

        //ver que este history es inventado, simula funciones, pero tiene los parametros fundamentales para que el componente funcione.
        // para ellos me debo fijar en el componente que parametros relacionas a history hay. 
        const history = {
            length : 10,
            push : jest.fn(),
            goBack : jest.fn(),
        }
        //cambio del shallow al mount por que como debo poner el useParam, debo envolver my componente en un higher order component

        const wrapper = mount (
            <MemoryRouter initialEntries={["/hero"]}>

                <HeroScreen history = {history}/>
              
            </MemoryRouter>
        
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("Redirect").exists()).toBe(true);
        
    });

    test('should shows an hero if the params exist and it is found ', () => {
  

        const wrapper = mount (
            <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
                <Route path = "/hero/:heroId" component = {HeroScreen} />
            </MemoryRouter>
        )

        expect (wrapper.find(".row").exists()).toBe(true)
    });

    test('should return at the back screen using PUSH', () => {
        const history = {
            length : 1,
            push : jest.fn(),
            goBack : jest.fn(),
        }
        
        const wrapper = mount (
            <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
                <Route
                 path = "/hero/:heroId"
                 component = {(props)=> <HeroScreen history = {history} />}
                 
                  />
            </MemoryRouter>
        )

        wrapper.find("button").prop("onClick")();

        expect(history.push).toHaveBeenCalledWith("/");
        expect(history.goBack).not.toHaveBeenCalled();   



    })

    
    test('should go further using goBaCK', () => {
        const history = {
            length : 10,
            push : jest.fn(),
            goBack : jest.fn(),
        }
        
        const wrapper = mount (
            <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
                <Route
                 path = "/hero/:heroId"
                 component = {(props)=> <HeroScreen history = {history} />}
                 
                  />
            </MemoryRouter>
        )

        wrapper.find("button").prop("onClick")();

        expect(history.goBack).toHaveBeenCalled();
  



    })

    test('should call Redirect', () => {
        const history = {
            length : 10,
            push : jest.fn(),
            goBack : jest.fn(),
        }
        
        const wrapper = mount (
            <MemoryRouter initialEntries={["/hero/marvel-spider11111111"]}>
                <Route
                 path = "/hero/:heroId"
                 component = {(props)=> <HeroScreen history = {history} />}
                 
                  />
            </MemoryRouter>
        )

        //si el heroe no existe, ese text. no deber√≠a de mostrarse nada
        expect(wrapper.text()).toBe("");
  



    })

    
     
    
    
})
