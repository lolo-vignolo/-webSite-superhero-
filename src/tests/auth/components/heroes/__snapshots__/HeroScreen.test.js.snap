// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test component <HeroScreen /> should shows itself correctly  1`] = `""`;
